### ISSUE TITLE
How to get docker-osx running on mac silicon (arm64)?

### ISSUE THREAD

**Issue by alanwilter** [sickcodes/Docker-OSX#547]
All I saw here was about running on Linux or Windows.

My idea was to have a simple macOS image with homebrew to build macOS related apps in a contained environment (Dockerfile with from ... and RUN brew install...), using my own Mac (silicon) as a host.

**Comment by WJacobsNL** [comment]
Here for the same reason :) Curious about the solution

**Comment by Pandaroses** [comment]
what

**Comment by danielgroen** [comment]
Same here!

**Comment by PATAPOsha** [comment]
Use `buildx` for building docker image.
E.g.:
`docker buildx build -t my-image --platform=linux/amd64 --progress=plain --build-arg DISPLAY=$DISPLAY .`
Then run container like so:
`docker run -it --platform linux/amd64 --net=host --name mi-container my-image`.

I tried Dockerfile starting with `FROM sickcodes/docker-osx:monterey`. It created and launched successfully.
However, I stuck on installing `homebrew` into container. Can not get how to add `homebrew` to the PATH.
There is no `~/.zprofile` or `~/.zshrc` o r`~/.profile` inside image...

**Comment by alanwilter** [comment]
Well, that did not quite work for me. Is your hosting Mac an M1/2 (silicon chip)?

When I run your instructions I got:

```bash
docker run --rm -it --platform linux/amd64 --net=host --name mi-container my-image
ssh-keygen: generating new host keys: RSA DSA ECDSA ED25519
nohup: appending output to 'nohup.out'
++ id -u
++ id -g
+ sudo chown 1000:1000 /dev/kvm
++ id -u
++ id -g
+ sudo chown -R 1000:1000 /dev/snd
+ [[ 4 = max ]]
+ [[ 4 = half ]]
++ id -u
++ id -g
+ sudo chown -R 1000:1000 /dev/snd
+ exec qemu-system-x86_64 -m 4000 -cpu Penryn,vendor=GenuineIntel,+invtsc,vmware-cpuid-freq=on,+ssse3,+sse4.2,+popcnt,+avx,+aes,+xsave,+xsaveopt,check, -machine q35,accel=kvm:tcg -smp 4,cores=4 -usb -device usb-kbd -device usb-tablet -device 'isa-applesmc,osk=ourhardworkbythesewordsguardedpleasedontsteal(c)AppleComputerInc' -drive if=pflash,format=raw,readonly=on,file=/home/arch/OSX-KVM/OVMF_CODE.fd -drive if=pflash,format=raw,file=/home/arch/OSX-KVM/OVMF_VARS-1024x768.fd -smbios type=2 -audiodev alsa,id=hda -device ich9-intel-hda -device hda-duplex,audiodev=hda -device ich9-ahci,id=sata -drive id=OpenCoreBoot,if=none,snapshot=on,format=qcow2,file=/home/arch/OSX-KVM/OpenCore/OpenCore.qcow2 -device ide-hd,bus=sata.2,drive=OpenCoreBoot -device ide-hd,bus=sata.3,drive=InstallMedia -drive id=InstallMedia,if=none,file=/home/arch/OSX-KVM/BaseSystem.img,format=qcow2 -drive id=MacHDD,if=none,file=/home/arch/OSX-KVM/mac_hdd_ng.img,format=qcow2 -device ide-hd,bus=sata.4,drive=MacHDD -netdev user,id=net0,hostfwd=tcp::10022-:22,hostfwd=tcp::5900-:5900, -device vmxnet3,netdev=net0,id=net0,mac=52:54:00:09:49:17 -monitor stdio -boot menu=on -vga vmware
kvm version too old
qemu-system-x86_64: failed to initialize kvm: Function not implemented
qemu-system-x86_64: falling back to tcg
QEMU 6.2.0 monitor - type 'help' for more information
(qemu) qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.avx [bit 28]
qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.80000007H:EDX.invtsc [bit 8]
qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.avx [bit 28]
qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.80000007H:EDX.invtsc [bit 8]
qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.avx [bit 28]
qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.80000007H:EDX.invtsc [bit 8]
qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.avx [bit 28]
qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.80000007H:EDX.invtsc [bit 8]
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_card_inum returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5701:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM default
alsa: Could not initialize DAC
alsa: Failed to open `default':
alsa: Reason: No such file or directory
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_card_inum returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5701:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM default
alsa: Could not initialize DAC
alsa: Failed to open `default':
alsa: Reason: No such file or directory
audio: Failed to create voice `dac'
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_card_inum returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5701:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM default
alsa: Could not initialize ADC
alsa: Failed to open `default':
alsa: Reason: No such file or directory
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_card_inum returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5701:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM default
alsa: Could not initialize ADC
alsa: Failed to open `default':
alsa: Reason: No such file or directory
audio: Failed to create voice `adc'
gtk initialization failed
```

But if I do:

```bash
run --rm -it --platform linux/amd64 --net=host --name mi-container --entrypoint=/bin/bash my-image
```

I got the prompt. However, `lscpu` is not auspicious:

```txt
[arch@docker-desktop OSX-KVM]$ lscpu
Architecture:           x86_64
  CPU op-mode(s):       32-bit
  Byte Order:           Little Endian
CPU(s):                 2
  On-line CPU(s) list:  0,1
Vendor ID:              0x00
  Model name:           -
    Model:              0
    Thread(s) per core: 1
    Core(s) per socket: 2
    Socket(s):          1
    Stepping:           0x0
    BogoMIPS:           48.00
    Flags:              fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm sb paca pacg dcpodp flagm2 frint
Vulnerabilities:
  Itlb multihit:        Not affected
  L1tf:                 Not affected
  Mds:                  Not affected
  Meltdown:             Not affected
  Mmio stale data:      Not affected
  Spec store bypass:    Vulnerable
  Spectre v1:           Mitigation; __user pointer sanitization
  Spectre v2:           Not affected
  Srbds:                Not affected
  Tsx async abort:      Not affected
```

I'm not surprised to `linux`, what I'm surprised is to see `x86_64`.

In the end, I'm hoping to see a `darwin` and `arm64`, like when I do `uname -a` in my host Mac (silicon).

I did try:

```bash
docker buildx build -t my-image --platform=darwin/amd64 --progress=plain --build-arg DISPLAY=$DISPLAY .
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 255B 0.0s done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/sickcodes/docker-osx:monterey
#3 DONE 0.8s

#4 [1/1] FROM docker.io/sickcodes/docker-osx:monterey@sha256:76289b8676b9c145f532d566bacd5700ac35edb1c8806ff65f3b6ffaf707bf7d
#4 resolve docker.io/sickcodes/docker-osx:monterey@sha256:76289b8676b9c145f532d566bacd5700ac35edb1c8806ff65f3b6ffaf707bf7d done
#4 CACHED

#5 exporting to image
#5 exporting layers done
#5 writing image sha256:33261377a6dd15c0ea1d77a3874a882540241bc69e1f6587288da8e7f368c88b done
#5 ERROR: operating system is not supported
------
 > exporting to image:
------
ERROR: failed to solve: operating system is not supported
```

So I'm back to square one, is there such image? Or, is there a way to build it?

**Comment by PATAPOsha** [comment]
> Well, that did not quite work for me. Is your hosting Mac an M1/2 (silicon chip)?

Yes, I'm using MacBook Pro m1 pro



> In the end, I'm hoping to see a `darwin` and `arm64`, like when I do `uname -a` in my host Mac (silicon).

From what I seen these "Docker-OSX" are based on linux `x86_64` image. So no, there is no clean `arm` macOs image. Correct me if I am wrong.

**Comment by patniemeyer** [comment]
This should definitely seems like it should be possible since UTM is just a GUI for running QEMU and it runs on M1 ARM.  (https://mac.getutm.app/).

**Comment by ylluminate** [comment]
@patniemeyer you can use QEMU actually with UTM to run pre-Apple Virtualization supported guest virtual machines. For example, Mojave and High Sierra can and should be run this way. We definitely need to see macOS support here since there are still essential usecases for older OS releases, especially Mojave.

**Comment by ylluminate** [comment]
Has anything happened on this front?

**Comment by AhirGhosh** [comment]
Hello all, any workable solution in this front please? Running on Apple M1 and getting this : 

`Status: Downloaded newer image for sickcodes/docker-osx:monterey
**WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested**
docker: Error response from daemon: error gathering device information while adding custom device "/dev/kvm": no such file or directory.`

**Comment by Destrocamil** [comment]
@AhirGhosh read the thread

**Comment by poyen-wu** [comment]
If someone is also trying to make this work on macOS because you want a macOS docker image, this isn't it. Docker-OSX is a docker image that sets up a QEMU VM in the container and runs macOS in the VM. (and on macOS that means you're running Linux version of QEMU in a Linux container since there's no macOS docker image) This is not a macOS docker image.

If you're really looking to run a macOS VM inside macOS, you can run a QEMU VM directly on macOS host, it will be faster and easier.

**Comment by ylluminate** [comment]
_So I ran this thread through gpt-5.2-pro-extended to get some closure on the various angles going on above. Hope this helps clarify anything that's outstanding:_

# Docker-OSX on Apple Silicon - State of the Union (March 2026)

This issue has been open since October 2022 with no maintainer response and no resolution in sight. Docker-OSX's last meaningful commit was November 2022, and it has 150+ open issues accumulating. The project works great for its original design target - x86_64 Linux hosts with KVM - but Apple Silicon support was never added and at this point never will be (unless forked). Here's the definitive picture as of March 2026.

## Why Docker-OSX can't work on Apple Silicon

Docker-OSX is **not a macOS container image**. It's a Linux container (x86_64) that bundles QEMU + OpenCore + macOS recovery media and boots a full x86_64 macOS VM inside the container, relying on **KVM** for hardware-accelerated virtualization on the Linux host.

On Apple Silicon Macs (M1 through M5):

- There is no x86_64 KVM. Apple's Hypervisor.framework only virtualizes arm64 guests.
- Forcing `--platform=linux/amd64` via Docker/Rosetta gets the Linux container layer running, but QEMU inside it falls back to **TCG** (pure software emulation) - which is the exact set of errors reported throughout this thread (`kvm version too old`, `failed to initialize kvm`, missing AVX, etc.).
- This results in a triple-nested emulation stack (macOS arm64 host → Docker Linux VM → amd64 container via Rosetta → QEMU TCG x86_64 macOS guest) that is effectively unusable for real work.
- **Nested KVM is not available** inside any Docker runtime on Apple Silicon. OrbStack's own docs explicitly state: *"It's not currently possible to support nested KVM virtualization on Apple Silicon. We're waiting for Apple to support it on the macOS side."* Docker Desktop deprecated its QEMU backend entirely in July 2025. So there's no trick involving OrbStack, Colima, or Docker Desktop that bypasses this as far as I can tell.

Also worth noting: `RUN brew install ...` in a Dockerfile executes in the **Linux image build layer**, not inside the macOS guest. For macOS Homebrew/Xcode you install inside the VM itself.

## "But doesn't Apple do containers now?"

Apple shipped a native [Containerization framework](https://github.com/apple/containerization) and [`container` CLI](https://github.com/apple/container) in macOS 26 Tahoe (WWDC 2025). It's OCI-compliant, open-source, Swift-based, and runs each Linux container in its own lightweight VM via Virtualization.framework. It's genuinely great for Linux container workflows on Mac.

But it only runs **Linux** containers. It cannot run macOS guests. It doesn't solve the problem in this thread.

## What actually works - by goal

### Goal 1: Contained/reproducible macOS build environment on Apple Silicon (modern macOS, fast)

**Use [Tart](https://github.com/cirruslabs/tart)** (actively maintained, v2.31.0 as of Feb 2026, 174 releases):

```bash
brew install cirruslabs/cli/tart
tart clone ghcr.io/cirruslabs/macos-tahoe-base:latest tahoe-base
tart run tahoe-base
```

Tart uses Apple's native Virtualization.framework for near-native arm64 performance. It has Docker-like CLI ergonomics (`tart clone`, `tart run`, `tart push/pull` to OCI registries), supports snapshots, shared directories, Packer integration, and was purpose-built for CI/dev workflows on Apple Silicon. Pre-built images are available for Monterey through Tahoe. This is the closest thing to "Docker for macOS VMs" that exists.

Alternatively, **UTM** in Apple Virtualization mode gives you a GUI-based experience with similar near-native performance for arm64 macOS guests (Monterey 12.4+).

### Goal 2: Older Intel macOS (Mojave / High Sierra / Catalina) on Apple Silicon

**Use [UTM](https://mac.getutm.app/) with its QEMU emulation backend.** This is the only viable free path.

- Apple's Virtualization.framework **blocks** pre-Ventura Intel macOS guests on M4/M5 chips entirely (black screen / instant failure). Even on M1-M3 it only goes back to Monterey.
- Parallels Desktop 26 has an x86 emulation "early technology preview" (Pro/Business/Enterprise only), but it only supports Windows and Linux x86_64 guests - **not macOS**. It's also limited to 1 vCPU, max 8GB RAM, no USB, and self-described as "really slow."
- VMware Fusion similarly does not support legacy Intel macOS guests on Apple Silicon.

UTM + QEMU is software emulation (TCG), so it's inherently slower than native hardware. But on M4 Max / M5 class hardware it's usable for legacy SDK work, testing, 32-bit app support, and old Xcode versions. Community configs from [adespoton/utmconfigs](https://github.com/adespoton/utmconfigs) (Mojave, High Sierra, Catalina, back to Mac OS 7) make setup straightforward.

This will never be "fast" compared to real Intel hardware. If speed is critical for sustained legacy workloads, a used Intel Mac Mini (~$150 on eBay) remains the most pragmatic option.

### Goal 3: Docker-OSX specifically, with proper performance

Run it on what it was built for: an **x86_64 Linux host with KVM**. Bare metal or cloud (Hetzner, OVH, etc.). That's the optimized path and always has been.

## Bottom line

The "macOS in a Docker container on Apple Silicon" dream is architecturally impossible without KVM, which Apple Silicon does not and will not expose for x86 guests. Docker-OSX is unmaintained for this use case. The ecosystem has moved on:

- **Modern macOS VM on Apple Silicon** → Tart (CLI/CI) or UTM (GUI)
- **Legacy Intel macOS on Apple Silicon** → UTM + QEMU emulation
- **Docker-OSX with real performance** → x86_64 Linux host with KVM

This issue should probably be closed with a note in the README pointing to Tart and UTM for Apple Silicon users - which the project FAQ already partially does.
